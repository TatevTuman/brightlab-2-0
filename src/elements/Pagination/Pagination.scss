@import "static/assets/styles/mixins";

.pagination {
  margin: 0 auto;
  width: fit-content;

  &-list {
    display: flex;
    align-items: center;
    margin: 0 -0.1rem;

    &-item {
      width: toRem(30);
      height: toRem(30);
      display: flex;
      align-items: center;
      justify-content: center;
      background: var(--primary);
      color: var(--white);
      border-radius: 4px;
      margin: 0 0.1rem;
      cursor: pointer;
      transition: transform 0.3s ease, background 0.2s ease;
      will-change: transform, background;

      &:focus {
        outline: none;
        color: var(--white);
      }

      &[data-active='true'], &:hover {
        outline: none;
        transform: scale(1.1);
        background: var(--black);
        color: var(--white);
      }
    }
  }
}
