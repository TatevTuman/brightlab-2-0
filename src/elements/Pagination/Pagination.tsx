import React from 'react'
import { PaginationResponse } from '@types'
import styles from './Pagination.module.scss'

interface PaginationProps extends PaginationResponse {
  limit?: number
  onPageClick(page: number): void
}

const Pagination: React.FC<PaginationProps> = props => {
  const { current, totalPages = 1, limit = 7, onPageClick } = props
  if (!current) return null

  /* Pages from the left and right to render */
  const pagesOffset = Math.floor(limit / 2)

  /* If current page is in page offset so current in first pages */
  const isFirstPages = current <= pagesOffset
  /* If current page + page offset greater than total pages the current is in last pages */
  const isLastPages = current + pagesOffset > totalPages!

  /* Start page */
  let start: number
  /* End page */
  let end: number

  if (isFirstPages) {
    /*
      If first pages:
        start: 1
        end: limit or total pages
    */
    start = 1
    end = Math.min(limit, totalPages!)
  } else if (isLastPages) {
    /*
      If last pages:
        start: total pages - limit without inclusion or 1
        end: total pages

      - 1 because of last page inclusion
    */

    start = Math.max(totalPages! - (limit - 1), 1)
    end = totalPages!
  } else {
    /*
      Else:
        start: current - page offset
        end: last page in a row or total pages
    */
    start = current - pagesOffset
    end = Math.min(current + pagesOffset, totalPages!)
  }

  const renderPages = () => {
    return new Array(limit).fill(0).map((_, index) => {
      const page = start + index

      /* End constrait */
      if (page > end) return

      const isActive = page === current

      return (
        <li
          key={index}
          tabIndex={0}
          className={styles.paginationListItem}
          onClick={() => onPageClick(page)}
          data-active={isActive}
        >
          {page}
        </li>
      )
    })
  }

  return (
    <nav role="pagination" className={styles.pagination}>
      <ul className={styles.paginationList}>{renderPages()}</ul>
    </nav>
  )
}

export default Pagination
