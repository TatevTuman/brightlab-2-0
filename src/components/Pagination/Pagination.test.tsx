import React from 'react'
import { render, cleanup } from '@testing-library/react'
import Pagination from './Pagination'
import userEvent from '@testing-library/user-event'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Pagination', () => {
  const props = {
    itemCount: 1000,
    total: 100,
    pageSize: 10,
    totalPages: 10,
    current: 1,
    limit: 5,
    onPageClick: jest.fn()
  }

  it('renders correctly', async () => {
    const { container, getByRole } = render(<Pagination {...props} />)

    expect(container).toMatchSnapshot()
    expect(getByRole('pagination')).toBeInTheDocument()
  })

  it('doesn`t render without current', async () => {
    const { container } = render(<Pagination {...props} current={0} />)
    const pagination = container.querySelector('nav[role="pagination"]')

    expect(pagination).toBeNull()
  })

  it('renders pages correctly', async () => {
    let current = 5
    let offset = current - Math.floor(current / 2)
    const { getByText, rerender } = render(<Pagination {...props} current={current} />)

    new Array(props.limit).fill(0).forEach((_, index) => {
      expect(getByText((index + offset).toString())).toBeInTheDocument()
    })

    expect(getByText('5')).toHaveAttribute('data-active', 'true')

    current = 9
    offset = current - Math.floor(current / 2) + 1 // 1 is inclusion
    rerender(<Pagination {...props} current={current} />)

    new Array(props.limit).fill(0).forEach((_, index) => {
      expect(getByText((index + offset).toString())).toBeInTheDocument()
    })

    expect(getByText('9')).toHaveAttribute('data-active', 'true')
  })

  it('renders pages with totalPages correctly', async () => {
    const current = 5
    const { getByText } = render(<Pagination {...props} current={current} totalPages={3} />)

    new Array(3).fill(0).forEach((_, index) => {
      expect(getByText((index + 1).toString())).toBeInTheDocument()
    })
  })

  it('clicks on page', async () => {
    const { getByText } = render(<Pagination {...props} />)

    userEvent.click(getByText('1'))
    expect(props.onPageClick).toHaveBeenCalledTimes(1)

    userEvent.click(getByText('2'))
    expect(props.onPageClick).toHaveBeenCalledTimes(2)
  })
})
