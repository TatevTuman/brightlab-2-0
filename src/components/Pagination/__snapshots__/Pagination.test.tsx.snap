// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div>
  <nav
    class="pagination"
    role="pagination"
  >
    <ul
      class="paginationList"
    >
      <li
        class="paginationListItem"
        data-active="true"
        tabindex="0"
      >
        1
      </li>
      <li
        class="paginationListItem"
        data-active="false"
        tabindex="0"
      >
        2
      </li>
      <li
        class="paginationListItem"
        data-active="false"
        tabindex="0"
      >
        3
      </li>
      <li
        class="paginationListItem"
        data-active="false"
        tabindex="0"
      >
        4
      </li>
      <li
        class="paginationListItem"
        data-active="false"
        tabindex="0"
      >
        5
      </li>
    </ul>
  </nav>
</div>
`;
